autoAdjustHwloc.hpp

```diff
@@ -23,154 +24,151 @@ class autoAdjust
+		for(size_t i=0; i < obj->arity; i++)
 		{
-			if (obj)
-			{
-				hwloc_cpuset_t cpuset;
-				/* Get a copy of its cpuset that we may modify. */
-				cpuset = hwloc_bitmap_dup(obj->cpuset);
-				size_t allcpu = hwloc_bitmap_to_ulong(cpuset);
-				/* Get only one logical processor (in case the core is
-				   SMT/hyperthreaded). */
-				hwloc_bitmap_singlify(cpuset);
+			if(obj->children[i]->type == type)
+				lambda(obj->children[i]);
+			else
+				findChildrenByType(obj->children[i], type, lambda);
+		}
+	}
 
+	inline bool isCacheExclusive(hwloc_obj_t obj)
+	{
+		const char* value = hwloc_obj_get_info_by_name(obj, "Inclusive");
+		return value != nullptr && value[0] == '1';
+	}
 
-				int firstNativeCore = hwloc_bitmap_first(cpuset);
+	// Top level cache isn't shared with other cores on the same package
+	// This will usually be 1 x L3, but can be 2 x L2 per package
+	void proccessTopLevelCache(hwloc_obj_t obj)
+	{
+		if(obj->attr == nullptr)
+			throw(std::runtime_error("Cache object hasn't got attributes."));
 
-				int nativeCores = hwloc_bitmap_weight(cpuset);
-				int numPus = obj->arity;
-				for (int i = 0; i < numPus && numHashes != 0 && firstNativeCore != -1; i++)
-				{
-					hwloc_obj_t pu = obj->children[i];
-					// only use native pu's
-					if (pu->type == HWLOC_OBJ_PU && hwloc_bitmap_isset( cpuset, i + firstNativeCore ))
-					{
-						// if no cache is available we give each native core a hash
-						int numUnit = numCachesLeft != 0 ? numCachesLeft : nativeCores;
-
-						// two hashes per native pu if number of hashes if larger than compute units
-						int power = numHashes > numUnit ? 2 : 1;
-						char strbuf[256];
-	
-						snprintf(strbuf, sizeof(strbuf), "   { \"low_power_mode\" : %s, \"no_prefetch\" : true, \"affine_to_cpu\" : %u },\n",
-							power == 2 ? "true" : "false", pu->os_index);
-						printer::inst()->print_str(strbuf);
-
-						// update number of free hashes
-						numHashes -= power;
-
-						// one cache is filled with hashes
-						if (numCachesLeft != 0) numCachesLeft--;
-					}
-				}
-			}
-		}
-		else
+		size_t PUs = 0;
+		findChildrenByType(obj, HWLOC_OBJ_PU, [&PUs](hwloc_obj_t found) { PUs++; } );
+
+		//Strange case, but we will handle it silently, surely there must be one PU somewhere?
+		if(PUs == 0)
+			return;
+
+		if(obj->attr->cache.size == 0)
 		{
-			for (int i = 0; i < obj->arity; i++)
-				getConfig(topology, obj->children[i], numHashes, numCachesLeft);
+			//We will always have one child if PUs > 0
+			if(obj->children[0]->type != HWLOC_OBJ_CACHE)
+				throw(std::runtime_error("The CPU doesn't seem to have a cache."));
+
+			//Try our luck with lower level caches
+			for(size_t i=0; i < obj->arity; i++)
+				proccessTopLevelCache(obj->children[i]);
+			return;
 		}
-	}
 
-	inline void findCache(hwloc_topology_t topology, hwloc_obj_t obj)
-	{
-		if (obj->type == HWLOC_OBJ_CACHE)
+		size_t cacheSize = obj->attr->cache.size;
+		if(!isCacheExclusive(obj))
```

Comment:
my fault this must be `isCacheExclusive(obj)`.
I always read this as is inclusive^^